include/statevector.hpp

```diff
@@ -0,0 +1,201 @@
+//////////////////////////////////////////////////////////////////////////////////////
+//
+// (C) Daniel Strano and the Qrack contributors 2017-2019. All rights reserved.
+//
+// This header defines buffers for Qrack::QFusion.
+// QFusion adds an optional "gate fusion" layer on top of a QEngine or QUnit.
+// Single bit gates are buffered in per-bit 2x2 complex matrices, to reduce the cost
+// of successive application of single bit gates to the same bit.
+//
+// Licensed under the GNU Lesser General Public License V3.
+// See LICENSE.md in the project root or https://www.gnu.org/licenses/lgpl-3.0.en.html
+// for details.
+
+#pragma once
+
+#include <algorithm>
+#include <map>
+#include <mutex>
+
+#include "common/qrack_types.hpp"
+
+namespace Qrack {
+
+class StateVector;
+class StateVectorArray;
+class StateVectorSparse;
+
+typedef std::shared_ptr<StateVector> StateVectorPtr;
+typedef std::shared_ptr<StateVectorArray> StateVectorArrayPtr;
+typedef std::shared_ptr<StateVectorSparse> StateVectorSparsePtr;
+
+// This is a buffer struct that's capable of representing controlled single bit gates and arithmetic, when subclassed.
+class StateVector {
+protected:
+    bitCapInt capacity;
+
+public:
+    StateVector(bitCapInt cap)
+        : capacity(cap)
+    {
+    }
+    virtual complex get(const bitCapInt& i) = 0;
+    virtual void set(const bitCapInt& i, const complex& c) = 0;
+    virtual void clear() = 0;
+    virtual void copy_in(const complex* inArray) = 0;
```

Comment:
Write and read are probably more common names

Reply:
Done.